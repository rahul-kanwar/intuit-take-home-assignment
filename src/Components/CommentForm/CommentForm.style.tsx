import styled from "styled-components";

export const Form = styled.form`
  display: flex;
  flex-direction: column;
  margin-bottom: 20px;
`;

export const Input = styled.input`
  margin-bottom: 10px;
  padding: 10px;
  font-size: 16px;
`;

export const TextArea = styled.textarea`
  margin-bottom: 10px;
  padding: 10px;
  font-size: 16px;
`;

export const Button = styled.button`
  padding: 10px;
  font-size: 16px;
  background-color: #007bff;
  color: white;
  border: none;
  cursor: pointer;
`;