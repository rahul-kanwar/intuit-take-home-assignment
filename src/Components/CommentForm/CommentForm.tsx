import React, { useState, ChangeEvent, FormEvent } from 'react';

import { createComment, Comment } from '../../api';
import { Form, Input, TextArea, Button } from './CommentForm.style';

interface CommentFormProps {
  onCommentAdded: (comment: Comment) => void;
}

const CommentForm: React.FC<CommentFormProps> = ({ onCommentAdded }) => {
  const [name, setName] = useState('');
  const [message, setMessage] = useState('');

  const handleSubmit = async (e: FormEvent) => {
    e.preventDefault();
    try {
      const newComment: Omit<Comment, 'id'> = {
        name,
        message,
        created: new Date().toISOString(),
      };
      const addedComment = await createComment(newComment);
      onCommentAdded({ ...addedComment, ...newComment });
      setName('');
      setMessage('');
    } catch (error) {
      console.error('Failed to add comment', error);
    }
  };

  return (
    <Form onSubmit={handleSubmit}>
      <Input
        type="text"
        placeholder="Name"
        value={name}
        onChange={(e: ChangeEvent<HTMLInputElement>) => setName(e.target.value)}
        required
      />
      <TextArea
        placeholder="Comment"
        value={message}
        rows={10}
        onChange={(e: ChangeEvent<HTMLTextAreaElement>) =>
          setMessage(e.target.value)
        }
        required
      />
      <Button type="submit">Add Comment</Button>
    </Form>
  );
};

export default CommentForm;
