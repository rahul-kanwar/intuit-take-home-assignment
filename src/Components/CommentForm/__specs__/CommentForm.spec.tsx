import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import CommentForm from '../CommentForm';
import 'jest-styled-components';

// Mock props
const mockOnCommentAdded = jest.fn();

describe('CommentForm Component', () => {
  it('renders correctly', () => {
    const { asFragment } = render(<CommentForm onCommentAdded={mockOnCommentAdded} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('submits the form with the correct data', () => {
    render(<CommentForm onCommentAdded={mockOnCommentAdded} />);

    fireEvent.change(screen.getByPlaceholderText('Name'), { target: { value: 'John' } });
    fireEvent.change(screen.getByPlaceholderText('Comment'), { target: { value: 'This is a test comment' } });

    fireEvent.click(screen.getByText('Add Comment'));

    expect(mockOnCommentAdded).toHaveBeenCalledWith(expect.objectContaining({
      name: 'John',
      message: 'This is a test comment',
    }));
  });

  it('resets input fields after submission', () => {
    render(<CommentForm onCommentAdded={mockOnCommentAdded} />);

    const nameInput = screen.getByPlaceholderText('Name');
    const messageInput = screen.getByPlaceholderText('Comment');

    fireEvent.change(nameInput, { target: { value: 'John' } });
    fireEvent.change(messageInput, { target: { value: 'This is a test comment' } });

    fireEvent.click(screen.getByText('Add Comment'));

    expect(nameInput.textContent).toBe('');
    expect(messageInput.textContent).toBe('');
  });
});
