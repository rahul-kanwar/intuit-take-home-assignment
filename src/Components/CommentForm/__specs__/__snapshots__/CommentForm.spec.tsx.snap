// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentForm Component renders correctly 1`] = `
<DocumentFragment>
  .c0 {
  display: flex;
  flex-direction: column;
  margin-bottom: 20px;
}

.c1 {
  margin-bottom: 10px;
  padding: 10px;
  font-size: 16px;
}

.c2 {
  margin-bottom: 10px;
  padding: 10px;
  font-size: 16px;
}

.c3 {
  padding: 10px;
  font-size: 16px;
  background-color: #007bff;
  color: white;
  border: none;
  cursor: pointer;
}

<form
    class="c0"
  >
    <input
      class="c1"
      placeholder="Name"
      required=""
      type="text"
      value=""
    />
    <textarea
      class="c2"
      placeholder="Comment"
      required=""
      rows="10"
    />
    <button
      class="c3"
      type="submit"
    >
      Add Comment
    </button>
  </form>
</DocumentFragment>
`;
